### ISSUE TITLE
Inductor gets stuck compiling code, unbounded RAM usage

### ISSUE THREAD

**Issue by mfbalin** [pytorch/pytorch#121197]
### 🐛 Describe the bug

When I run a GNN workload with a specific dataset, the compilation process gets stuck on step 2.

### Error logs

```
[2024-03-05 04:00:37,435] [0/0] torch.fx.experimental.symbolic_shapes: [INFO] create_env
[2024-03-05 04:00:37,668] [0/0] torch.fx.experimental.symbolic_shapes: [INFO] create_symbol s0 = 341523 for L['x'].size()[0] [2, 9223372036854775806]
[2024-03-05 04:00:37,669] [0/0] torch.fx.experimental.symbolic_shapes: [INFO] create_symbol s1 = 128 for L['x'].size()[1] [2, 9223372036854775806]
[2024-03-05 04:00:37,853] [0/0] torch.fx.experimental.symbolic_shapes: [INFO] create_symbol s2 = 471405 for L['subgraphs'][0].sampled_csc.indices.size()[0] [2, 9223372036854775806]
[2024-03-05 04:00:37,856] [0/0] torch.fx.experimental.symbolic_shapes: [INFO] create_symbol s3 = 95082 for L['subgraphs'][0].sampled_csc.indptr.size()[0] [2, 9223372036854775806]
[2024-03-05 04:00:37,884] [0/0] torch.fx.experimental.symbolic_shapes: [INFO] eval s3 - 1 <= s0 [guard added] at ocalscratch/dgl-3/examples/sampling/graphbolt/lightning/../../pyg/node_classification_3.py:82 in convert_to_pyg (_decomp/decompositions.py:756 in slice_forward)
[2024-03-05 04:00:37,951] [0/0] torch.fx.experimental.symbolic_shapes: [INFO] eval Ne(s3 - 1, 1) [guard added] at torch_geometric/utils/_scatter.py:80 in scatter (_refs/__init__.py:394 in _broadcast_shapes)
[2024-03-05 04:00:37,951] [0/0] torch.fx.experimental.symbolic_shapes: [INFO] eval Ne(s3 - 1, 1) [guard added] at torch_geometric/utils/_scatter.py:80 in scatter (_refs/__init__.py:400 in _broadcast_shapes)
[2024-03-05 04:00:37,996] [0/0] torch.fx.experimental.symbolic_shapes: [INFO] set_replacement s1 = 128
[2024-03-05 04:00:37,996] [0/0] torch.fx.experimental.symbolic_shapes: [INFO] set_replacement s1 = 128
[2024-03-05 04:00:37,996] [0/0] torch.fx.experimental.symbolic_shapes: [INFO] eval Eq(s1, 128) [guard added] at torch_geometric/nn/dense/linear.py:147 in forward (_meta_registrations.py:1980 in meta_mm)
[2024-03-05 04:00:38,017] [0/0] torch.fx.experimental.symbolic_shapes: [INFO] create_symbol s4 = 124196 for L['subgraphs'][1].sampled_csc.indices.size()[0] [2, 9223372036854775806]
[2024-03-05 04:00:38,020] [0/0] torch.fx.experimental.symbolic_shapes: [INFO] create_symbol s5 = 12758 for L['subgraphs'][1].sampled_csc.indptr.size()[0] [2, 9223372036854775806]
[2024-03-05 04:00:38,039] [0/0] torch.fx.experimental.symbolic_shapes: [INFO] eval s5 - 1 <= s3 - 1 [guard added] at ocalscratch/dgl-3/examples/sampling/graphbolt/lightning/../../pyg/node_classification_3.py:82 in convert_to_pyg (_decomp/decompositions.py:756 in slice_forward)
[2024-03-05 04:00:38,098] [0/0] torch.fx.experimental.symbolic_shapes: [INFO] eval Ne(s5 - 1, 1) [guard added] at torch_geometric/utils/_scatter.py:80 in scatter (_refs/__init__.py:394 in _broadcast_shapes)
[2024-03-05 04:00:38,099] [0/0] torch.fx.experimental.symbolic_shapes: [INFO] eval Ne(s5 - 1, 1) [guard added] at torch_geometric/utils/_scatter.py:80 in scatter (_refs/__init__.py:400 in _broadcast_shapes)
[2024-03-05 04:00:38,143] [0/0] torch.fx.experimental.symbolic_shapes: [INFO] create_symbol s6 = 12033 for L['subgraphs'][2].sampled_csc.indices.size()[0] [2, 9223372036854775806]
[2024-03-05 04:00:38,147] [0/0] torch.fx.experimental.symbolic_shapes: [INFO] create_symbol s7 = 1025 for L['subgraphs'][2].sampled_csc.indptr.size()[0] [2, 9223372036854775806]
[2024-03-05 04:00:38,165] [0/0] torch.fx.experimental.symbolic_shapes: [INFO] eval s7 - 1 <= s5 - 1 [guard added] at ocalscratch/dgl-3/examples/sampling/graphbolt/lightning/../../pyg/node_classification_3.py:82 in convert_to_pyg (_decomp/decompositions.py:756 in slice_forward)
[2024-03-05 04:00:38,229] [0/0] torch.fx.experimental.symbolic_shapes: [INFO] eval Ne(s7 - 1, 1) [guard added] at torch_geometric/utils/_scatter.py:80 in scatter (_refs/__init__.py:394 in _broadcast_shapes)
[2024-03-05 04:00:38,230] [0/0] torch.fx.experimental.symbolic_shapes: [INFO] eval Ne(s7 - 1, 1) [guard added] at torch_geometric/utils/_scatter.py:80 in scatter (_refs/__init__.py:400 in _broadcast_shapes)
[2024-03-05 04:00:38,277] [0/0] torch._dynamo.symbolic_convert: [INFO] Step 1: torchdynamo done tracing forward (RETURN_VALUE)
[2024-03-05 04:00:38,283] [0/0] torch._dynamo.output_graph: [INFO] Step 2: calling compiler function inductor
```

### Minified repro

Running with the environment variable gets stuck too, producing no output.

The code in question is here: https://github.com/dmlc/dgl/pull/7183

It can be run with `python examples/sampling/pyg/node_classification_3.py --dataset=ogbn-papers100M`.

After building and installing DGL from source.

### Versions

<details><summary>Details</summary>
<p>

Collecting environment information...
PyTorch version: 2.3.0a0+ebedce2
Is debug build: False
CUDA used to build PyTorch: 12.3
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.3 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: Could not collect
CMake version: version 3.28.1
Libc version: glibc-2.35

Python version: 3.10.12 (main, Nov 20 2023, 15:14:05) [GCC 11.4.0] (64-bit runtime)
Python platform: Linux-5.15.0-88-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 12.3.107
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: NVIDIA A100-SXM4-80GB
GPU 1: NVIDIA A100-SXM4-80GB
GPU 2: NVIDIA A100-SXM4-80GB
GPU 3: NVIDIA A100-SXM4-80GB

Nvidia driver version: 545.23.08
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.9.0.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv.so.9.0.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn.so.9.0.0
/usr/lib/x86_64-linux-gnu/libcudnn_engines_precompiled.so.9.0.0
/usr/lib/x86_64-linux-gnu/libcudnn_engines_runtime_compiled.so.9.0.0
/usr/lib/x86_64-linux-gnu/libcudnn_graph.so.9.0.0
/usr/lib/x86_64-linux-gnu/libcudnn_heuristic.so.9.0.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops.so.9.0.0
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Address sizes:                      43 bits physical, 48 bits virtual
Byte Order:                         Little Endian
CPU(s):                             256
On-line CPU(s) list:                0-255
Vendor ID:                          AuthenticAMD
Model name:                         AMD EPYC 7702 64-Core Processor
CPU family:                         23
Model:                              49
Thread(s) per core:                 2
Core(s) per socket:                 64
Socket(s):                          2
Stepping:                           0
Frequency boost:                    enabled
CPU max MHz:                        2183.5930
CPU min MHz:                        1500.0000
BogoMIPS:                           4000.17
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid aperfmperf rapl pni pclmulqdq monitor ssse3 fma cx16 sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 hw_pstate ssbd mba ibrs ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 cqm rdt_a rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local clzero irperf xsaveerptr rdpru wbnoinvd amd_ppin arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif v_spec_ctrl umip rdpid overflow_recov succor smca sme sev sev_es
Virtualization:                     AMD-V
L1d cache:                          4 MiB (128 instances)
L1i cache:                          4 MiB (128 instances)
L2 cache:                           64 MiB (128 instances)
L3 cache:                           512 MiB (32 instances)
NUMA node(s):                       2
NUMA node0 CPU(s):                  0-63,128-191
NUMA node1 CPU(s):                  64-127,192-255
Vulnerability Gather data sampling: Not affected
Vulnerability Itlb multihit:        Not affected
Vulnerability L1tf:                 Not affected
Vulnerability Mds:                  Not affected
Vulnerability Meltdown:             Not affected
Vulnerability Mmio stale data:      Not affected
Vulnerability Retbleed:             Mitigation; untrained return thunk; SMT enabled with STIBP protection
Vulnerability Spec rstack overflow: Mitigation; safe RET
Vulnerability Spec store bypass:    Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:           Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:           Mitigation; Retpolines, IBPB conditional, STIBP always-on, RSB filling, PBRSB-eIBRS Not affected
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Not affected

Versions of relevant libraries:
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.24.4
[pip3] onnx==1.15.0rc2
[pip3] optree==0.10.0
[pip3] pytorch-quantization==2.1.2
[pip3] torch==2.3.0a0+ebedce2
[pip3] torch_geometric==2.5.0
[pip3] torch-tensorrt==2.3.0a0
[pip3] torchdata==0.7.1a0
[pip3] torchmetrics==1.3.1
[pip3] torchtext==0.17.0a0
[pip3] torchvision==0.18.0a0
[pip3] triton==2.2.0+e28a256
[conda] Could not collect

</p>
</details>

**Comment by ezyang** [comment]
Can you conveniently use py-spy to get a backtrace from the stuck process?

**Comment by mfbalin** [comment]
Output of `py-spy`, I killed the process after a while of it getting stuck. ![profile](https://github.com/pytorch/pytorch/assets/16190101/7f40b923-ddcc-49dd-8780-f76ae4df9fb7)

**Comment by ezyang** [comment]
I need to actually see the stack trace. You could do py-spy dump to get it, or if you post the actual svg that will help too

**Comment by mfbalin** [comment]
> I need to actually see the stack trace. You could do py-spy dump to get it, or if you post the actual svg that will help too

The actual svg file is below, not sure which command to use to get the dump.

https://github.com/pytorch/pytorch/assets/16190101/bb916bd8-7e3a-415a-9f0f-7862a422decb

**Comment by mfbalin** [comment]
@ezyang Pasted the output above, looks like a lot of processes are getting started and there might be some kind of deadlock somewhere.

**Comment by ezyang** [comment]
Yeah, it looks deadlocked for no obvious reason. Will need to investigate more I guess :/

**Comment by ezyang** [comment]
Runs repro for CPU OOM. Tmux crashes.

![image](https://github.com/pytorch/pytorch/assets/13564/e672c920-d0d0-419a-85b9-6d37d368e527)

**Comment by mfbalin** [comment]
@ezyang When I use a different dataset, this issue does not show itself so it such a weird bug. The following on another dataset works just fine.

`python dgl/examples/sampling/graphbolt/pyg/node_classification_advanced.py --dataset=ogbn-products --torch-compile`

**Comment by ezyang** [comment]
gdb suggests that you are forking the process at some point, and all the forks are now stuck. If you can figure out who is forking, make it stop doing that (use spawn or something).

**Comment by ezyang** [comment]
Or maybe it is us that is forking...

**Comment by mfbalin** [comment]
> Or maybe it is us that is forking...

I should not be forking in this training script, even the dataloader must be initialized with 0 workers.

**Comment by ezyang** [comment]
So, I am currently seeing what happens if we `import torch._dynamo` before doing any of the other imports in the example script. For Reasons(TM) PyTorch spins up some forks when you import torch._dynamo, and I suspect that a different import is initializing CUDA and then hilarity ensues.

**Comment by ezyang** [comment]
This has not seem to have worked but maybe that's just because I need to import something else maybe

**Comment by ezyang** [comment]
OK I was wrong, torch._dynamo doesn't work, but `TORCHINDUCTOR_COMPILE_THREADS=1` does. Can you see if that also resolves the problem on your end too?

**Comment by ezyang** [comment]
I think what I want to do is detect if there are threads before we fork for torch.compile and either error or disable multithreaded compilation when this occurs. But this is complicated by the fact that OpenBLAS spawns threads by default, but it registers an atfork handler to shutdown the threads gracefully, so OpenBLAS's threads "shouldn't count".

What's kind of nifty, though, is that you can explicitly ask OpenBLAS to shut down its threads. Need to find a better way of discovering the library, but `ctypes.CDLL("/data/users/ezyang/a/pytorch-env/lib/python3.10/site-packages/numpy.libs/libopenblas64_p-r0-0cf96a72.3.23.dev.so").blas_thread_shutdown_()` does it locally for me.

So concretely, here is what I propose we do here: before we fork compilation threads, on Linux only, we:

1. Test if OpenBLAS is loaded (somehow). If it is, call `blas_thread_shutdown_()`
2. Look at /proc/PID/tasks to see how many threads there are
3. If there is more than 1 thread, we emit a warning and disable multithreaded compilation
4. Otherwise, we now fork

WDYT?

cc @malfet

**Comment by jansel** [comment]
IMO it might make sense to just stop using multiprocessing here and directly Popen a subprocess we control.

**Comment by zdevito** [comment]
I agree with using subprocess. We did try this before. There are some meta-specific things where multiprocessing is better supported, but I think it is worth fixing those rather than trying to use multiprocessing which adds so many additional issues.

**Comment by malfet** [comment]
Well, at least on Mac, as soon as one imports Torch there are more than 1 thread, see this example:
```
% python -c "import torch;from ctypes import byref, c_long, c_int, CDLL;h=CDLL('libSystem.dylib');tarr, tcnt=c_long(0),c_int(-1);h.task_threads(h.mach_task_self(), byref(tarr), byref(tcnt));print(tcnt)"
c_int(8)
```

**Comment by mfbalin** [comment]
> OK I was wrong, torch._dynamo doesn't work, but `TORCHINDUCTOR_COMPILE_THREADS=1` does. Can you see if that also resolves the problem on your end too?

I confirm that it resolved the problem on my end as well.

**Comment by ezyang** [comment]
Well, I know how to turn off forks when it is unsafe, but I don't know how to "fix the subprocess/multiprocess situation in Meta". Perfect enemy of good?

**Comment by masnesral** [comment]
@mfbalin , the subprocess-based approach mentioned by @jansel above is now the default on recent nightlies if you'd like to give it a try. (i.e., you shouldn't need the `TORCHINDUCTOR_COMPILE_THREADS=1` workaround)

**Comment by mfbalin** [comment]
I was importing `import torch._inductor.codecache` at the beginning of my code but deleting it seems to work now.
